server/src/main/kotlin/org/javacs/kt/externalsources/ClassPathSourceArchiveProvider.kt

```diff
@@ -0,0 +1,29 @@
+package org.javacs.kt.externalsources
+
+import org.javacs.kt.CompilerClassPath
+import java.io.File
+import java.nio.file.Path
+import java.nio.file.Paths
+
+class ClassPathSourceArchiveProvider(
+    private val cp: CompilerClassPath
+) : SourceArchiveProvider {
+    override fun fetchSourceArchive(compiledArchive: Path): Path? =
+        getJdkSource(compiledArchive) ?: cp.classPath.firstOrNull { it.compiledJar == compiledArchive }?.sourceJar
+
+    /**
+     * Checks if the given path is inside the JDK. If it is, we return the corresponding source zip.
+     * Note that this method currently doesn't take into the account the JDK version, which means JDK source code
+     * is only available for JDK 9+ builds.
+     * TODO: improve this resolution logic to work for older JDK versions as well.
+     */
+    private fun getJdkSource(path: Path): Path? {
+        cp.javaHome?.let {
+            val javaHomePath = File(it).toPath()
+            if (path == javaHomePath) {
+                return Paths.get(path.toString(), "lib", "src.zip")
+            }
+        }
+        return null
+    }
+}
```

Comment:
Could we move the JDK source resolution to a separate `SourceArchiveProvider` (something along the lines of `JdkSourceArchiveProvider`) and have this one only handle the class path-based resolution?

Reply:
Well, we could but I think it would require bigger changes to the content provider logic, since the current `ClassContentProvider` relies on a single `SourceArchiveProvider` (and we need both the `ClassPathSourceArchiveProvider` and the `JdkSourceArchiveProvider`)

Reply:
I made some changes. Let me know if that works for you. I basically used the same strategy we have for the classpath resolvers, which works well and it's pretty clean.

Reply:
Looks good!